### ISSUE TITLE
Task Server Protocol

### ISSUE THREAD

**Issue by domenkozar** [cachix/devenv#1457]
# Task Server Protocol

https://devenv.sh allows defining tasks in your favorite language using [JSON-RPC](https://www.jsonrpc.org/specification#response_object)
protocol and exposing them as an executable :

```nix
{
  task.serverProtocol = [ "myexecutable" ];
}
```

## Listing tasks

Which would launch `myexecutable /tmp/rando.sock` and `devenv` would on startup immediately ask for a list of tasks:

```json
{
  "jsonrpc": "2.0",
  "method": "initialize",
  "params": {},
  "id": 1
}
```

And the server responds:

```json
{
  "jsonrpc": "2.0",
  "result": {
    "tasks": [{
      "name": "prefix:custom",
      "depends": []
    }]
  },
  "id": 1
}
```

## Running tasks

Then the client can ask the server to run a task:

```json
{
  "jsonrpc": "2.0",
  "method": "run",
  "params": {
    "task": "prefix:name",
    "inputs": {},
    "outputs": {}
  },
  "id": 2
}
```

And server streams updates about stdout/stderr:

```json
{
  "jsonrpc": "2.0",
  "method": "log",
  "params": {
    "task": "prefix:name",
    "line": "some text",
    "stderr": false,
    "time": "20240828T212611.11Z",
  }
}
```

Until the final response from the server with outputs and the final status of the task.

```json
{
  "jsonrpc": "2.0",
  "result": {
    "task": "prefix:name",
    "outputs": {},
    "status": "succeeded"
  },
  "id": 2
}
```

## Implementation

- Use [reth-ipc](https://github.com/paradigmxyz/reth/tree/main/crates/rpc/ipc) for the Client using (jsonrspee)[https://github.com/paritytech/jsonrpsee/)
- A test suite with a defined set of tasks that Server SDK should implement for the test suite to run against
- Rust SDK for the Server

**Comment by bobvanderlinden** [comment]
I was a bit surprised by this. I thought the tasks were usually short-lived operations that just run one or more commands in a certain order and then just end.

The TSP seems to indicate this is a long lived process that can interactively start one or more tasks. It almost seems like the protocol could be used for the `processes` of `devenv up`?

That's pretty cool, I'm just wondering what the intention is.

Just to illustrate, the same could be done by `myexecutable run task < json_file`? Where myexecutable always lives as long as the tasks it's running.

That said, the methods/responses make sense to me. I just wasn't sure where the tsp would get its configuration from if that resides in devenv.nix?

Nitpick (feel free to ignore): The stderr being bool was a bit surprising, but the alternatives I thought up might not be better (`io: stdout` or even `fileNo: 1`).

**Comment by domenkozar** [comment]
@bobvanderlinden you're correct to spot the intention for tasks to be an implementation detail for processes. That way we get systemd-level dependency trees also for processes.

> Just to illustrate, the same could be done by myexecutable run task < json_file? Where myexecutable always lives as long as the tasks it's running.

There are two issues here when I was thinking about the design:

- providing language specific interface via command line parsing is inflexible for future changes
- we need a way for myexecutable to report logging, etc at which point you're already introducing bidirectional channel

> I just wasn't sure where the tsp would get its configuration from if that resides in devenv.nix?

it gets configuration from initialization phase that happens as soon as executable runs.

> Nitpick (feel free to ignore): The stderr being bool was a bit surprising, but the alternatives I thought up might not be better (io: stdout or even fileNo: 1).

I had the same thought, the alternative would be `line_stdout` and `line_stderr`?

**Comment by domenkozar** [comment]
See https://github.com/cachix/devenv/issues/1471

**Comment by bennyandresen** [comment]
Very interested in this! I've built a PoC for Clojure (with a GraalVM target, for startup speed). Maybe this is a way to marry babashka tasks (common in the Clojure world) and making them available to devenv. So a "devenv up" would just work(tm) for Clojure devs.

**Comment by domenkozar** [comment]
@bennyandresen happy to code up the TSP if you'd like to take this forward and test!

**Comment by bennyandresen** [comment]
Yeah, very much so.

Just thinking out loud for Clojure/babashka integration, in case others are also interested in this.
We often use babashka as a task runner for Clojure projects, except the tasks or processes are written in Clojure.
But things like standing up a PostgresQL DB and filling it with initial data is done in other ways. That's where Just or in my case devenv comes in.

So in my mind, based on my PoC given this description one could use metadata inside a `bb.edn` (the babashka project file that makes babashka a task runner) so that they could be exported to devenv via the TSP.

As an example, a normal `bb.edn` file looks like this:
```clj
{:tasks
 {test
  {:doc "clojure -Srepro -M:poly test <args>"
   :task (apply poly "test" *command-line-args*)}

  test-all-verbose
  {:doc "bb test :all :verbose"
   :task (poly "test" :all :verbose)}

  outdated
  {:doc "clojure -Srepro -M:outdated"
   :task (clojure-repro "-M:outdated")}}}
```

you could then add `^:devenv` or other metadata to a key like `test-all-verbose` (= `^:devenv test-all-verbose`) which then makes it so it gets picked up by the TSP and it can be used further downstream.

My imagined use case would be to export the processes inside `bb.edn` that one needs, like the clojure backend process or processes in combination with the devenv provided redis and postgres for example by merely setting a metadata inside `bb.edn` which doesn't affect babashka usage but is a value-add for devenv users.

**Comment by bennyandresen** [comment]
> @bennyandresen if having the client implementation would allow you to go further with this, I'm happy to write up a PoC
Yes, that would help. I'm quite excited about this integration.

**Comment by justinTM** [comment]
hi @domenkozar could i bother you to ELI5? in our use case, we have gitlab+aws and use terraform, docker images, lambda functions, and python/bash scripts. if we are either running tasks on local dev machines or via CI/CD, would this have much benefit? it's not like we have a build server to communicate with over a network.

i'm only a few weeks into using devenv, but i feel like any invocation of a python script via a devenv task, script, or process would be simpler than writing a whole RPC server to execute the tasks...?

**Comment by domenkozar** [comment]
@justinTM it's mostly for exposing tasks from existing code bases, like you could write all your tasks in Python.

The idea is that you write Server implementation in a language once and then it should be trivial to add new tasks!

I've been prototyping this in past few days and I hope to have a PR up soon :)

**Comment by justinTM** [comment]
thank you for your reply @domenkozar , really amazing work you and the team are doing here!
